lms/djangoapps/course_api/blocks/transformers/milestones.py

```diff
@@ -35,22 +45,39 @@ def collect(cls, block_structure):
         block_structure.request_xblock_fields('is_proctored_enabled')
         block_structure.request_xblock_fields('is_practice_exam')
         block_structure.request_xblock_fields('is_timed_exam')
+        block_structure.request_xblock_fields('entrance_exam_id')
+
+    def transform(self, usage_info, block_structure):
+        """
+        Modify block structure according to the behavior of milestones and special exams.
+        """
 
-    def transform_block_filters(self, usage_info, block_structure):
-        if usage_info.has_staff_access:
-            return [block_structure.create_universal_filter()]
+        course_key = block_structure.root_block_usage_key.course_key
+        user_can_skip = EntranceExamConfiguration.user_can_skip_entrance_exam(usage_info.user, course_key)
+        required_content = milestones_helpers.get_required_content(course_key, usage_info.user)
 
         def user_gated_from_block(block_key):
             """
             Checks whether the user is gated from accessing this block, first via special exams,
             then via a general milestones check.
             """
-            return (
-                settings.FEATURES.get('ENABLE_SPECIAL_EXAMS', False) and
-                self.is_special_exam(block_key, block_structure)
-            ) or self.has_pending_milestones_for_user(block_key, usage_info)
+            if usage_info.has_staff_access:
+                return False
+            elif self.has_pending_milestones_for_user(block_key, usage_info):
+                return True
+            elif self.gated_by_required_content(block_key, block_structure, user_can_skip, required_content):
+                return True
+            elif (settings.FEATURES.get('ENABLE_SPECIAL_EXAMS', False) and
+                  (self.is_special_exam(block_key, block_structure) and
+                   not self.include_special_exams)):
+                return True
+            return False
 
-        return [block_structure.create_removal_filter(user_gated_from_block)]
+        for block_key in block_structure.topological_traversal():
+            if user_gated_from_block(block_key):
+                block_structure.remove_block(block_key, False)
+            else:
+                self.add_special_exam_info(block_key, block_structure, usage_info)
```

Comment:
@dianakhuang @nasthagiri: My very imperfect understanding was that the collect phase added data and the transform phase mostly pruned.  However, it is called "transform" and not "prune".  Can one of you help me understand what should happen where, and why and if `add_special_exam_info` makes sense during transform?  I admit I did not dig into code, I am just going by names and previous understanding.
: Is this the right place to add this info based on your intentions of collect vs transform?